src/rez/build_process.py

```diff
@@ -119,14 +136,15 @@ def __init__(self, working_dir, build_system, package=None, vcs=None,
                                            self.package.config.build_directory)
 
     @property
-    def package(self):
+    def package(self) -> DeveloperPackage:
         return self.build_system.package
 
     @property
-    def working_dir(self):
+    def working_dir(self) -> str:
         return self.build_system.working_dir
 
-    def build(self, install_path=None, clean=False, install=False, variants=None):
+    def build(self, install_path: str | None = None, clean: bool = False,
```

Comment:
Using `str | None` means we need to drop support for python 3.7. I'm not sure we are ready for this yet.

Reply:
No, use of `str | None` is find in python 3.7 as long as you use `from __future__ import annotations`.  This behavior ensures that type annotations are recorded as strings within `__annotations__` attributes, which means they are not evaluated at runtime unless [`inspect.get_annoations`](https://docs.python.org/3/library/inspect.html#inspect.get_annotations) is called. 

The only thing breaking python 3.7 compatibility here is the use of `TypedDict` and `Protocol`, as mentioned in [another comment](https://github.com/AcademySoftwareFoundation/rez/pull/1761#issuecomment-2119317337).  I presented 3 options for preserving the use of these classes in the other comment.  

I noticed that the only python 3.7 tests that are currently run are for MacOS, which I took as an indicator that python 3.7 would be dropped soon.   Is there a schedule for deprecation?

Reply:
By the way, I fixed the python 3.7 compatibility issue with `TypedDict` and `Protocol`, so that should not be a blocker anymore.